### PR TITLE
chore: avoid unnecessary allocations

### PR DESCRIPTION
**Pull request by DaniPopes**
Split off from #558

And some other minor clean-ups

### CODE REVIEW

File: crates/primitives/src/specification.rs [hunk 1]
```diff
@@ -1,9 +1,10 @@
+#![allow(non_camel_case_types)]
```
Comment (DaniPopes) [inline crates/primitives/src/specification.rs:1]: `rust-analyzer` gives a warning for me on the struct without this attribute at the module level

File: crates/primitives/src/bytecode.rs [hunk 2]
```diff
@@ -44,16 +47,18 @@ pub struct Bytecode {
...
+    #[inline]
     fn default() -> Self {
         Bytecode::new()
     }
 }
 
 impl Bytecode {
-    /// Create [`Bytecode`] with one STOP opcode.
+    /// Creates a new [`Bytecode`] with exactly one STOP opcode.
+    #[inline]
```
Comment (rakita) [inline crates/primitives/src/bytecode.rs:58]: The compiler is smart and in most cases will do a good job on the inline front, so adding `#[inline]` everywhere is just noisy.

File: crates/primitives/src/bytecode.rs [hunk 3]
```diff
@@ -75,11 +82,15 @@ impl Bytecode {
...
 
-    /// Create new raw Bytecode with hash
+    /// Creates a new raw Bytecode with the given hash.
     ///
     /// # Safety
-    /// Hash need to be appropriate keccak256 over bytecode.
+    ///
+    /// The given `hash` has to be equal to the [`keccak256`] hash of the given `bytecode`.
+    #[inline]
+    #[cfg_attr(debug_assertions, track_caller)]
```
Comment (rakita) [inline crates/primitives/src/bytecode.rs:91]: Why is this added, in essence, `unsafe` is here for a reason user of this should be careful what he does

Comment (DaniPopes) [reply crates/primitives/src/bytecode.rs:91]: Right, it's just an extra check in debug mode. I can remove it if you want

File: crates/primitives/src/bytecode.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/primitives/src/bytecode.rs:58]: This is not the case for public functions, which will not get inlined across crate boundaries without LTO: <https://nnethercote.github.io/perf-book/inlining.html>

Also `#[inline]` is just a hint, and given that these functions are very trivial I don't see why not to hint to the compiler that they should be inline (like `Vec::len`).

I've had this discussion before, I don't want to argue about this again so if you're really keen on this I'll just remove all of the hints.

Comment (rakita) [reply crates/primitives/src/bytecode.rs:58]: I was thinking just about in crate use, I tried to play with it with the opcode function and would have better results when I just leave it to the compiler. Let's leave it with `#[inline]` as it makes sense for `Bytecode`.

File: crates/primitives/src/bytecode.rs (hunk 3 above)
Comment (rakita) [reply crates/primitives/src/bytecode.rs:91]: Let us remove it, I would assume there is a possibility for one to intentionally switch the bytecode of a different hash.
